idea-plugin/src/main/kotlin/arrow/meta/ide/plugins/quotes/cache/QuoteCacheService.kt

```diff
@@ -0,0 +1,48 @@
+package arrow.meta.ide.plugins.quotes.cache
+
+import arrow.meta.ide.dsl.application.services.Id
+import arrow.meta.ide.dsl.application.services.IdService
+import org.jetbrains.kotlin.descriptors.DeclarationDescriptor
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.psi.KtFile
+import java.util.concurrent.ConcurrentHashMap
+
+/**
+ * a QuoteCache implementation as a IdService, which stores Quote transformations on disk.
+ * This is an initial implementation, as it does not populate the index of other connected projects,
+ * and the cache stored on disk get out of sync with external changes to the disk, projects sharing source files or other situations.
+ * fixme cache both per module? modules may define different ktFiles for the same package fqName
+ * TODO: register this via Meta.
+ */
+private object QuoteCacheService : IdService<ConcurrentHashMap<KtFile, QuoteInfo>>, QuoteCache {
+  override var value: Id<ConcurrentHashMap<KtFile, QuoteInfo>> =
+    Id.just(ConcurrentHashMap<KtFile, QuoteInfo>())
+
+  override val size: Int
+    get() = value.extract().size
+
+  override fun packages(): List<FqName> =
+    value.extract().values.map { it.first }.distinct()
+
+  override fun descriptors(packageFqName: FqName): List<DeclarationDescriptor> =
+    value.extract().values.filter { it.first == packageFqName }.map { it.second }.flatten()
+
+  override fun removeQuotedFile(file: KtFile): QuoteInfo? =
+    value.extract().remove(file).also { _ ->
```

Comment:
isn't this calling `remove(file)` twice on the same instance of the ConcurrentHashMap? `map` returns a copy of the `Id`, but both reference the same instance of the HashMap.

Reply:
Nop, the first remove is a pure computation that won't affect the underlying data structure. the second is impure in the sense that it uses the impure map function from `Id<A>` transforms the data structure here ConcurrentHashMap and `removeQuotedFile` results with the output of the pure computation above. 

It's ugly because morphisms defined used in `Id<A>` need to be pure and `map` returns the new data structure, whereas in this case for  `removeQuotedFile` is interested in the new mapping to other descriptors. 
I am totally open to change the algebra in `QuoteCache`, but we should open a new Issue for that.